src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs

```diff
@@ -199,7 +194,14 @@ public void Tick()
 
             // It's safe to use UtcNowUnsynchronized since Tick is called by the Heartbeat.
             var now = _systemClock.UtcNowUnsynchronized;
-            _timeoutControl.Tick(now);
+
+            lock (_streams)
```

Comment:
Taking a lock in our Heartbeat on every HTTP/3 connection every tick and looping over a dictionary really scares me. We should look into kicking off the locked looping logic on another thread like we do for HTTP/2.

Reply:
Where is that done?

Reply:
We discussed this offline, but for other potential readers [this call to Input.CancelPenddingRead()](https://github.com/dotnet/aspnetcore/blob/228d8de71700abf6f92da3c6734878b2d3d8a2ad/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs#L1114) effectively dispatches to the receive loop where we check for timeouts.